pkg/sql/update.go

```diff
@@ -422,17 +426,31 @@ func (ss scalarSlot) checkColumnTypes(row []tree.TypedExpr) error {
 //
 // The row buffer is modified in-place with the result of the
 // checks.
-func enforceLocalColumnConstraints(row tree.Datums, cols []sqlbase.ColumnDescriptor) error {
+func enforceLocalColumnConstraints(
+	ctx *tree.EvalContext, row tree.Datums, cols []sqlbase.ColumnDescriptor,
+) error {
 	for i := range cols {
 		col := &cols[i]
 		if !col.Nullable && row[i] == tree.DNull {
 			return sqlbase.NewNonNullViolationError(col.Name)
 		}
-		outVal, err := sqlbase.LimitValueWidth(&col.Type, row[i], &col.Name)
+		var err error
+		if !col.Type.Equivalent(row[i].ResolvedType()) {
+			if _, ok := tree.FindCast(
+				row[i].ResolvedType().Oid(),
+				col.Type.Oid(),
+				tree.CastContextAssignment,
+			); ok {
+				row[i], err = tree.PerformCast(ctx, row[i], &col.Type)
+				if err != nil {
+					return err
+				}
+			}
+		}
```

Comment:
Determining to cast or not during execution won't work because computed column values must be derived from the casted value. This is currently causing bugs like below in CRDB:

```
defaultdb> create table t (c char, s string as (c) stored);
CREATE TABLE

defaultdb> insert into t(c) values (' '::STRING);
INSERT 1

defaultdb> select length(c), length(s) from t;
  length | length
---------+---------
       0 |      1
(1 row)
```

Instead we'll have to wrap mutation inputs with assignment cast expressions in optbuilder if the types are not equivalent. These assignment casts could be a new syntax like `ASSIGNMENT_CAST(x AS INT)` or `CAST(x AS INT ASSIGNMENT)`.

Or we could use a pattern introduced in #70115 to pass types to a builtin function. Something like `crdb_internal.assignment_cast(x, NULL::INT)`.